### PR TITLE
Url encode to escape special characters

### PR DESCRIPTION
**Pull request by MehulG**
# Description
Implemented a new mod `str encode-url` to encode the string to escape special characters as done by browsers.

# User Facing changes
For a user to convert `https://example.com/foo bar` into `https://example.com/foo%20bar`, he would want to run the following command in nu:
```bash
> "https://example.com/foo bar" | str encode-url
https://example.com/foo%20bar
```
This also works for strings in list
```bash
> ['https://example.com/foo bar', 'https://example.com/a>b', '中文字/eng/12 34'] | str encode-url
╭───┬─────────────────────────────────────────╮
│ 0 │ https://example.com/foo%20bar           │
│ 1 │ https://example.com/a%3Eb               │
│ 2 │ %E4%B8%AD%E6%96%87%E5%AD%97/eng/12%2034 │
╰───┴─────────────────────────────────────────╯
```
# Related Issue
#7496 

@rgwood

### CODE REVIEW

**Comment by fdncred** [comment]
I think this looks like a great addition to nushell as it's been asked for many times. However, I'm not sure the name of the command is right. When people search nushell for a command to do this, I'm not sure they'll think of `str encode-url`.

My first thought would be to name it `url encode` since we already have a `url` base command. My second thought would be to make it `encode url`. What do other folks think?

**Comment by MehulG** [comment]
My thought behind this naming was this might not only be used for urls but for removing escape characters from any string as in the third example of the list

**Comment by rgwood** [comment]
Thanks for the PR @MehulG! I'll try to make time to take a close look at it soon.

@fdncred yeah, naming is hard :sweat_smile:. I think we should avoid `encode url`, @webbedspace had a good point [here](https://github.com/nushell/nushell/issues/7496#issuecomment-1356340980):

> the meaning of "encode" used by `encode` is something completely different to what you mean here. `encode` is a string -> bytes operation that takes a string (which has no "inherent" encoding) and turns it into the byte representation for that given Unicode encoding.

`url encode` or `str url-encode` both sound OK to me.

**Comment by fdncred** [comment]
My comment focused on the discoverability and what seems most intuitive for users. As long as `url` is in the name we'll be able to find it with `help -f url`. I get the point about encoding other strings, but it makes the most sense to me to use `url encode` since this is what most libraries and websites call this functionality.

I won't lose any sleep over calling it anything. renaming commands is easy. so, I won't fight over the name. I'm happy to land it as-is or change it to what I prefer, either is good to me.

**Comment by rgwood** [comment]
OK, let's go with `url encode`.

File: crates/nu-command/Cargo.toml [hunk 1]
```diff
@@ -88,6 +88,7 @@ titlecase = "2.0.0"
 toml = "0.5.8"
 unicode-segmentation = "1.8.0"
 url = "2.2.1"
+url-escape = "0.1.1"
```
Comment (rgwood) [inline crates/nu-command/Cargo.toml:91]: Is there a reason for choosing the [`url-escape`](https://lib.rs/crates/url-escape) crate?

It's not widely used (~2000 downloads/month); did you consider [`percent-encoding`](https://lib.rs/crates/percent-encoding) or [`urlencoding`](https://lib.rs/crates/urlencoding)
 instead?

Comment (MehulG) [reply crates/nu-command/Cargo.toml:91]: Ah, yes you are right. will change it to use [`urlencoding`](https://lib.rs/crates/urlencoding)

**Comment by MehulG** [comment]
I was wondering should we escape charterers for `/` and `:` as well? it would produce a bit of confusing results for users to do so;
` https://example.com/foo bar` would get encoded to `https%3A%2F%2Fexample%2Ecom%2Fa%3Eb` for example

**Comment by aidanharris** [comment]
> I was wondering should we escape charterers for `/` , `.` and `:` as well? it would produce a bit of confusing results for users to do so; ` https://example.com/foo bar` would get encoded to `https%3A%2F%2Fexample%2Ecom%2Fa%3Eb` for example Excluding the above all alphanumeric chars would be url encoded.

What about having two subcommands `url encode-uri` and `url encode-uri-component` and using the same behaviour as JavaScript? That would satisfy both use cases:

```
> encodeURI('https://example.com/foo bar')
'https://example.com/foo%20bar'
> encodeURIComponent('https://example.com/foo bar')
'https%3A%2F%2Fexample.com%2Ffoo%20bar'
```

File: crates/nu-command/Cargo.toml (hunk 1 above)
Comment (sholderbach) [reply crates/nu-command/Cargo.toml]: Mhh so if I grok it correctly from the docs the `url-escape` crate supports maintaining the path components while the others encode full urls to be used as a component/query?

Comment (MehulG) [reply crates/nu-command/Cargo.toml]: yep, that's true @sholderbach . But @rgwood does make a good point about it not being widely used. Instead with any of the other 2 libraries mentioned, we can customize the code anyway as I did with latest code in this PR.

**Comment by fdncred** [comment]
> I was wondering should we escape charterers ...

I was thinking that having a flag that would allow uri-component similar to what @aidanharris said but doing it in one command instead of having two commands. So, the default would be equivalent to encodeURI and the flag would allow similar functionality of encodeURIComponent. Thoughts?

**Comment by rgwood** [comment]
Hah, I was just writing up a reply with the same thoughts as @fdncred:

> I was wondering should we escape charterers for `/` , `.` and `:` as well?

I agree that this would confuse many users. We should probably support both approaches, but let's use a flag (`--all`?) instead of creating 2 separate commands.

**Comment by MehulG** [comment]
Ah yes! a `all` flag seems pretty intuitive. Will implement that :)

**Comment by MehulG** [comment]
Just added the `--all` flag, could you guys review the feature?

**Comment by fdncred** [comment]
I tested the `-a` and other use cases. Seems to work well. Good job! I also changed the usage text a bit.

**Review by sholderbach — APPROVED** [review]
Looks good to me! Thanks for contributing to nushell!

**Comment by fdncred** [comment]
I should've probably changed `web safe` to `web-safe`. @sholderbach's `non-alphanumeric` change made me think of it, ugh.

**Pull request merged**